cmd/update-controller/main.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/coreos/bootkube/pkg/cluster"
+	"github.com/coreos/bootkube/pkg/cluster/components"
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+func main() {
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	run(clientset.NewForConfigOrDie(config))
+}
+
+func run(client clientset.Interface) {
+	glog.Info("update controller running")
+	versionChan := make(chan *components.Version)
+	handleConfigChange := func(newConfigMap *v1.ConfigMap) {
+		newVersion, err := parseNewVersion(newConfigMap)
+		if err != nil {
+			glog.Error(err)
+			return
+		}
+		versionChan <- newVersion
+	}
+	updateCallback := func(_, newObj interface{}) {
+		glog.Info("begin update callback")
+		newConfigMap, ok := newObj.(*v1.ConfigMap)
+		if !ok {
+			glog.Infof("Wrong type for update callback, expected *v1.ConfigMap, got: %T", newObj)
+			return
+		}
+		handleConfigChange(newConfigMap)
+	}
+	addCallback := func(obj interface{}) {
+		glog.Info("begin add callback")
+		newConfigMap, ok := obj.(*v1.ConfigMap)
+		if !ok {
+			glog.Infof("Wrong type for update callback, expected *v1.ConfigMap, got: %T", obj)
+			return
+		}
+		handleConfigChange(newConfigMap)
+	}
+	cu, err := cluster.NewClusterUpdater(client)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	go func(client clientset.Interface, updater *cluster.ClusterUpdater) {
+		for v := range versionChan {
```

Comment:
I feel like queuing up version changes could be problematic. Ideally we are only targeting the "currently set" version -- not what might have been set and happens to be queued.

Let's start with just a single process targeting a single version, and periodically re-running (and checking desired version each time). As follow up in functionality, we should allow an upgrade to be stopped mid-process rather than having to wait to completion to reset / change desired.